lib/propolis/src/vmm/time.rs

```diff
@@ -0,0 +1,788 @@
+    // timer values, it is critical to set this value prior to importing device
+    // state such that existing timers are normalized correctly. As with booting
+    // a guest, on the target, it should be set to the hrtime of the host when
+    // the guest would have booted, had it booted on that target.
+    //
+    // An example may be helpful. Consider a guest that has 5 days of uptime,
+    // booted on a host with 30 days of uptime. Suppose that guest is migrated
+    // with a device timer that should fire 1 second in the future.
+    //
+    // +=================================================================+
+    // | hrtime (source) | guest hrtime | boot_hrtime  | timer value     |
+    // +-----------------------------------------------------------------+
+    // | 30 days         | 5 days       |(30 - 5) days | src hrtime + 1s |
+    // |                 |              | 25 days      | 30 days + 1s    |
+    // +=================================================================+
+    //
+    // Suppose the guest is then migrated to a host with 100 days of uptime.
+    // On migration, the existing timer is normalized before export by
+    // subtracting out boot_hrtime:
+    //       normalized = timer - boot_hrtime
+    //                  = (30 days + 1 sec) - 25 days
+    //                  = 5 days + 1 sec
+    //
+    // When the timer is imported, it is denormalized by adding back in
+    // the new boot_hrtime. The timer should still fire 1 second from the
+    // current hrtime of the host. The target hrtime is 100 days, so the timer
+    // should fire at 100 days + 1 sec.
+    //
+    // Working backwards to get the new boot_hrtime, we have:
+    //
+    //       denormalized = normalized + boot_hrtime
+    //       boot_hrtime  = denormalized - normalized
+    //       boot_hrtime  = (100 days + 1 sec) - (5 days + 1 sec)
+    //       boot_hrtime  = 95 days
+    //
+    // And on the target, the timer should still fire 1 second into the future
+    // as expected:
+    //
+    // +=====================================================================+
+    // | hrtime (target) | guest hrtime | boot_hrtime   | timer value        |
+    // +---------------------------------------------------------------------+
+    // | 100 days        | 5 days       |(100 - 5) days |     5 days + 1 sec |
+    // |                 |              | 95 days       |   + 95 days        |
+    // |                 |              |               | = 100 days + 1 sec |
+    // +=====================================================================+
+    //
+    // NB: It is possible for boot_hrtime to be negative; this occurs if a
+    // guest has a longer uptime than its host (an expected common case for
+    // migration). This is okay: hrtime is a signed value, and the normalization
+    // maths still work with negative values.
+    //
+
+    // guest_uptime_ns  = source hrtime - source boot_hrtime
+    let guest_uptime_ns: i64 = src
```

Comment:
I'm having trouble squaring this with the comment on line 181. If the exported timer value is normalized, hasn't the source boot_hrtime been subtracted from src.hrtime already?

Reply:
We chatted about this offline -- there are no timers being normalized/denormalized here; this is part of the calculation of the new boot_hrtime. This code and comment could make what is happening here more clear.